features/poller/shell/src/main/java/org/opennms/netmgt/filters/shell/FilterCommand.java

```diff
@@ -0,0 +1,116 @@
+import org.opennms.netmgt.model.OnmsNode;
+import org.springframework.transaction.support.TransactionCallback;
+import org.springframework.transaction.support.TransactionOperations;
+
+@Command(scope = "filters", name = "filter", description = "Enumerates nodes/interfaces that match a give filter")
+@Service
+public class FilterCommand implements Action {
+
+    @Reference
+    private FilterDao filterDao;
+
+    @Reference
+    private NodeDao nodeDao;
+
+    @Reference
+    private TransactionOperations transactionOperations;
+
+    @Argument(description = "A filter Rule", required = true, multiValued = false)
+    private String filterRule;
+
+    @Override
+    public Object execute() throws Exception {
+        transactionOperations.execute((TransactionCallback<Void>) status -> {
+            boolean matching = false;
+            SortedMap<Integer, String> nodeMap = null;
+            List<InetAddress> ipAddressList = null;
+            try {
+                matching = filterDao.isRuleMatching(filterRule);
+                if (matching) {
+                    nodeMap = filterDao.getNodeMap(filterRule);
+                    ipAddressList = filterDao.getActiveIPAddressList(filterRule);
+                }
+            } catch (Exception e) {
+                // pass
+            }
+            if (!matching) {
+                System.out.printf("No matching nodes/interfaces for this rule.\n");
+                return null;
+            }
+            List<OnmsNode> nodes = new ArrayList<>();
+            if (nodeMap != null && !nodeMap.isEmpty()) {
+                nodeMap.keySet().forEach(nodeId -> nodes.add(nodeDao.get(nodeId)));
+                System.out.printf("---Matching nodes---\n");
+            }
+            for (OnmsNode node : nodes) {
+                List<String> categoryNames = new ArrayList<>();
+
+                node.getCategories().forEach(category -> categoryNames.add(category.getName()));
+                String categories = node.getCategories().isEmpty() ? ""
+                        : "categories=" + categoryNames.stream().collect(Collectors.joining(","));
+                String foreignId = (node.getForeignId() != null) ? "" : "foreignId=" + node.getForeignId();
+                String foreignSource = (node.getForeignSource() != null) ? "" : "foreignSource=" + node.getForeignSource();
+                String location = nodeDao.getLocationForId(node.getId());
+                System.out.printf("nodeId=%d nodeLabel=%s location=%s %s %s %s \n", node.getId(), node.getLabel(),
+                        location, foreignId, foreignSource, categories);
+
+            }
+            List<String> ipAddresses = new ArrayList<>();
+            if (ipAddressList != null && !ipAddressList.isEmpty()) {
+                ipAddressList.stream().forEach(ipAddress -> ipAddresses.add(ipAddress.getHostAddress()));
+                System.out.printf("---Matching IP Addresses that are active---\n");
```

Comment:
What do you think about showing the nodes followed by the associated IP addresses on the node that were matched rather than displaying these separately?

Reply:
Makes sense.  Updating.